public/js/comms.js

```diff
@@ -1,33 +1,49 @@
-// This comms will be loaded into each client, will use the socket.io library
-
-// comms.js
-function Comms (roomId, endpointId) {
-  this.endpointid = endpointId;
-  this.roomid = roomId;
-  this.handlers = {};
-  Comms.endpoints[this.endpointid] = this;
-  this.ws = io(); // global io() function made available via web sockets
-  var self = this;
-  this.ws.on('message', function (msg) {
-    msg = JSON.parse(msg);
-    self.handlers[msg.app + '.' + msg.method](this, msg.from, msg.params);
-  });
-}
+// Main comms class
+class Comms {
+
+  // Takes endopint and room ID
+  constructor (roomId, endpointId) {
+    var self = this;
+    this.endpointid = endpointId;
+    this.externalendpoints = new Set();
+    this.roomid = roomId;
+    this.handlers = {};
+    this.ws = io();
+
+    this.ws.on('message', function (msg) {
+      msg = JSON.parse(msg);
+      // Simply calls the handler with 'from' and 'params' (not self)
+      self.handlers[msg.app + '.' + msg.method](msg.sender, msg.params);
+    });
 
-Comms.endpoints = {}; // Holds endpoints
+    Comms.endpoints[this.endpointid] = this;
+  }
 
-Comms.getEndPointID = function (endpointId) { // Get endpoints
-  return Comms.endpoints[endpointId];
-};
+  // Register Handler
+  registerHandler (app, method, cb) {
+    this.handlers[`${app}.${method}`] = cb;
+  }
+  // Add external endpoint
+  addExternalEndpoint (endpoint) {
+    this.externalendpoints.add(endpoint);
+  }
+
+  // Send Method
+  send (app, method, receiver, params) {
```

Comment:
It might be worth giving `params` a default value of `''`? I notice that elsewhere, you end up having to manually enter an empty string every time